gax-java/gax/src/test/java/com/google/api/gax/batching/FlowControllerTest.java

```diff
@@ -596,19 +597,20 @@ public void testNumberOfBytesOutOfBoundaryWontDeadlock() throws Exception {
     // will be blocked by reserve 10
     Thread t =
         new Thread(
-            new Runnable() {
-              @Override
-              public void run() {
-                try {
-                  flowController.reserve(0, 100);
-                } catch (FlowControlException e) {
-                }
+            () -> {
+              try {
+                flowController.reserve(0, 100);
+              } catch (FlowControlException e) {
+                throw new AssertionError(e);
               }
             });
     t.start();
+
     // wait for thread to start, and check it should be blocked
-    Thread.sleep(50);
-    assertEquals(State.WAITING, t.getState());
+    assertByPollingEvery(10, TimeUnit.MILLISECONDS)
+        .withTimeout(200, TimeUnit.MILLISECONDS)
+        .thatEventually(() -> assertEquals(State.WAITING, t.getState()));
```

Comment:
I do feel introducing new assertion class is overkill (unnecessary complexity). Doesn't this "while" loop below suffice the purpose?

```
    int attempt = 0;
    do {
      Thread.sleep(10);
      assertTrue(
          "Waited 200 milliseconds but the thread does not become WAITING state", attempt++ < 20);
    } while (State.WAITING != t.getState());
```

Reply:
Perhaps different schools of thought on how test code should be written. In this case, I didn't want to add more inlined logic to an already-complex test case, saw I needed this logic at least twice, and thought it likely that the other flakes we're working on might find it useful.

But I do agree that inlining the above has a certain low-edit benefit to it, so I will defer to whatever you and Blake prefer.

Reply:
I don't have a strong opinion on this as I can see advantages on both sides. I agree with you if we just want to retry a failing unit test, as we have other ways for it like using [surefire plugin](https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html) to retry failed unit tests. But it does provide more granularity control for flaky tests and it could be easier for us to understand the flaky pattern for each test.

Reply:
On a separate note, it does not solve this problem, but Junit 5 has things like [RepeatedTest](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/RepeatedTest.html) that could be useful for us.

Reply:
On continuing to think over this, I'd still prefer not to inline this logic. Putting the retry logic in a class means it can be tested to ensure it's working as expected.

Reply:
The complexity in creating AssertByPolling.java is that we have to add its test. In the inline code, I don't feel that simple logic needs test.

Reply:
Discussed with @suztomo and implemented agreement: Simplified assertion.